### PR TITLE
Brian Zhou EQUIIS QA Test

### PR DESCRIPTION
**Pull request by briandesigns**
# **Testing method**

**Domain package**
contains mostly data state classes with setter and getters for every field. Contains little to no logic therefore does not require testing

**Data package**
contains data access logic as well as repositories. This is an important package as it contains the backend of this (mostly) CRUD application. To ensure data integrity, we need to make sure that the CRUD operations are performed correctly therefore all datastore and repository classes should be thoroughly tested for their correctness. I have written the test class for DatabaseFlashcardDataStore only, however the unit tests for the other classes in the package should look almost identical. Note in my test cases, I only tested for the most general cases in order to save time. To be thorough, I would need to tests the corner cases such as attempting to retrieve element from an empty list, providing extreme values as arguments and checking for a methods runtime etc. 

**Generator package**
should be unit tested the exact same way as the important classes in Data package.

**App package**
This package contains the android UI classes such as Activities, fragments, adapters and presenters etc. The best way to test these classes is to use instrumentation tests that run with a given context. We can use these tests to check whether an UI element such as a buttons and views are actually rendered during runtime. We can also check if action listeners are being triggered by events and if the corresponding action is being taken. It would be a good idea to test the classes listed above in details to make sure that no UI element is broken as the success of android applications are heavily affected by faultless and fluid UI. I have no experience in writing instrumentation tests so I have not written any. However If I did, I would put a lot of focus on the presenters as they are the classes that directly calls the CRUD operations in the Data package.

**Intra-app integration testing**
To test if the app brings me the correct UI element such as a different activity or an animation playback etc, I could use a UI testing framework like Espresso. For example, When I click on the play button of a lesson, I want to know that I will enter the flashcard lesson mode and nothing else. 

**Inter-app integration testing**
This app doesn't seem to interact much with other applications other than maybe Google Play for ratings and perhaps Messenger for recommending this app to your friends so I wouldn't focus too much on inter-app integration testing. However if I want to be absolutely thorough I would use a testing framework like UI Automator.

### CODE REVIEW

File: app/src/test/java/com/dbychkov/data/datastore/DatabaseFlashcardDataStoreTest.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    /**
+     * init a new databaseflashcardatastore
+     *
+     * @throws Exception
+     */
+    @After
+    public void tearDown() throws Exception {
+        databaseFlashcardDataStore = new DatabaseFlashcardDataStore(mock(DaoSession.class));
```
Comment (naerymdan) [inline app/src/test/java/com/dbychkov/data/datastore/DatabaseFlashcardDataStoreTest.java:45]: Why would you init a new DatabaseFlashcardDataStore instance in the teardown method?

File: app/src/test/java/com/dbychkov/data/datastore/DatabaseFlashcardDataStoreTest.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        assertEquals(flashcardEntity, result);
+    }
+
+    @Test
+    public void testRemoveFlashcard() throws Exception {
+        FlashcardEntity flashcardEntity = new FlashcardEntity(1L);
+        databaseFlashcardDataStore.addFlashcard(flashcardEntity);
+        FlashcardEntity result =
+                databaseFlashcardDataStore.getFlashcardById(1L).toBlocking().singleOrDefault(null);
+        if (result == null) {
```
Comment (naerymdan) [inline app/src/test/java/com/dbychkov/data/datastore/DatabaseFlashcardDataStoreTest.java:64]: Wouldn't that be better served with a

"assertNotEqual(result, null);" instead of the if statement?

File: gradle/wrapper/gradle-wrapper.properties [hunk 3]
```diff
@@ -1,6 +1,6 @@
-#Tue Jan 12 13:39:55 MSK 2016
+#Fri May 06 11:55:15 EDT 2016
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-2.10-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-2.10-all.zip
```
Comment (naerymdan) [inline gradle/wrapper/gradle-wrapper.properties:6]: Why graddle-2.10-all instead of the -bin one?

File: app/src/test/java/com/dbychkov/data/datastore/DatabaseFlashcardDataStoreTest.java (hunk 1 above)
Comment (briandesigns) [inline app/src/test/java/com/dbychkov/data/datastore/DatabaseFlashcardDataStoreTest.java:45]: the goal was to de-allocated the memory of the databaseFlashcardDataStore that is being used in all of the tests. Now that I think about it, its completely redundant to implement the teardown method for this test class specifically since setUp() already takes care of it at the start of every test. I think I was simply trying to be thorough without paying too much attention to the implications.

File: app/src/test/java/com/dbychkov/data/datastore/DatabaseFlashcardDataStoreTest.java (hunk 2 above)
Comment (briandesigns) [inline app/src/test/java/com/dbychkov/data/datastore/DatabaseFlashcardDataStoreTest.java:64]: You are right, an assertNotEqual statement would be much better here. The reason why I used an if statement at the time was because using a fail() allowed me to provide a customized failure message for debugging purposes. I've never used fail() before and just wanted to see how it works.

File: gradle/wrapper/gradle-wrapper.properties (hunk 3 above)
Comment (briandesigns) [inline gradle/wrapper/gradle-wrapper.properties:6]: I updated to Android Studio 2.+ before taking on this project who had an older Android Studio version. I had to perform a series of updates and syncs including gradle version in order to achieve compatibility and this line must have been one of the changes it made.

**Comment by naerymdan** [comment]
Thanks!
